### PR TITLE
ARROW-17297: [Java][Doc] Adding documentation to interact between C++ to Java via C Data Interface

### PR DESCRIPTION
**Pull request by davisusanibar**
Adding documentation to interact between C++ to Java via C Data Interface for:
- ImportArray
- ImportRecordBatch

### CODE REVIEW

**Comment by lidavidm** [comment]
Can you create a JIRA? This isn't minor

File: docs/source/java/cdata.rst [hunk 1]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+        }
+    }
+
+Compile our Java code to generate our jar with all dependencies needed for.
+
+.. code-block:: shell
+
+    $ mvn clean install
+    $ cp target/cpptojava-1.0-SNAPSHOT-jar-with-dependencies.jar cpptojava.jar
+    $ cp cpptojava.jar <c++_project_path>
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:363]:
```suggestion
    $ cp target/cpptojava-1.0-SNAPSHOT-jar-with-dependencies.jar <c++ project path>/cpptojava.jar
```

File: docs/source/java/cdata.rst [hunk 2]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+            return root;
+        }
+
+        public static void main(String[] args) {
+            populateFieldVectorToExport();
+            populateVectorSchemaRootToExport();
+        }
+    }
+
+Compile our Java code to generate our jar with all dependencies needed for.
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:357]:
```suggestion
Build the JAR and copy it to the C++ project.
```

File: docs/source/java/cdata.rst [hunk 3]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+            intVector.allocateNew(3);
+            intVector.setSafe(0, 1);
+            intVector.setSafe(1, 2);
+            intVector.setSafe(2, 3);
+            intVector.setValueCount(3);
+            System.out.println("[Java - side]: FieldVector: \n" + intVector);
+            return intVector;
+        }
+
+        public static void fillvectorschemarootfromjavatocpp(long schema_id, long array_id){
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:333]: Can we name these appropriately?

File: docs/source/java/cdata.rst [hunk 4]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+    import org.apache.arrow.vector.FieldVector;
+    import org.apache.arrow.vector.IntVector;
+    import org.apache.arrow.vector.VectorSchemaRoot;
+
+    import java.util.Arrays;
+
+    public class ToBeCalledByCpp {
+        final static BufferAllocator allocator = new RootAllocator();
+
+        public static void fillfieldvectorfromjavatocpp(long schema_id, long array_id){
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:315]: Can we follow Java naming conventions? `schemaAddress`

File: docs/source/java/cdata.rst [hunk 5]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+        }
+
+        public static FieldVector populateFieldVectorToExport(){
+            IntVector intVector = new IntVector("int-to-export", allocator);
+            intVector.allocateNew(3);
+            intVector.setSafe(0, 1);
+            intVector.setSafe(1, 2);
+            intVector.setSafe(2, 3);
+            intVector.setValueCount(3);
+            System.out.println("[Java - side]: FieldVector: \n" + intVector);
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:329]:
```suggestion
            System.out.println("[Java] FieldVector: \n" + intVector);
```

File: docs/source/java/cdata.rst [hunk 6]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+
+.. code-block:: shell
+
+    $ mvn clean install
+    $ cp target/cpptojava-1.0-SNAPSHOT-jar-with-dependencies.jar cpptojava.jar
+    $ cp cpptojava.jar <c++_project_path>
+
+**C++ Side**
+
+This process fetch Java JVM references to call methods needed for, via the C Data Interface:
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:367]:
```suggestion
This application uses JNI to call Java code, but transfers data (zero-copy) via the C Data Interface instead.
```

File: docs/source/java/cdata.rst [hunk 7]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+
+.. code-block:: cpp
+
+    #include <iostream>
+    #include <arrow/api.h>
+    #include <arrow/c/bridge.h>
+    #include <jni.h>
+
+    JNIEnv* create_vm(JavaVM ** jvm) {
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:376]:
```suggestion
    JNIEnv* CreateVM(JavaVM** jvm) {
```

File: docs/source/java/cdata.rst [hunk 8]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+        JNIEnv *env;
+        JavaVMInitArgs vm_args;
+        JavaVMOption options[2];
+        options[0].optionString = "-Djava.class.path=cpptojava.jar"; // java jar name
+        options[1].optionString = "-DXcheck:jni:pedantic";
+        vm_args.version = JNI_VERSION_1_8;
+        vm_args.nOptions = 2;
+        vm_args.options = options;
+        int status = JNI_CreateJavaVM(jvm, (void**) &env, &vm_args);
+        if (status < 0) printf("\n<<<<< Unable to Launch JVM >>>>>\n");
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:386]: std::abort instead?

File: docs/source/java/cdata.rst [hunk 9]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+        return env;
+    }
+
+    int main() {
+        JNIEnv* env;
+        JavaVM* jvm;
+        env = create_vm(&jvm);
+        if (env == NULL) return 1;
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:394]:
```suggestion
        if (env == nullptr) return 1;
```

File: docs/source/java/cdata.rst [hunk 10]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+        jclass  javaClassToBeCalledByCpp = NULL;
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:395]:
```suggestion
        jclass javaClassToBeCalledByCpp = NULL;
```

File: docs/source/java/cdata.rst [hunk 11]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+        javaClassToBeCalledByCpp = env ->FindClass("ToBeCalledByCpp");
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:396]: Can you make sure to format the source code?

File: docs/source/java/cdata.rst [hunk 12]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+        if ( javaClassToBeCalledByCpp != NULL ) {
+            jmethodID fillfieldvectorfromjavatocpp = NULL;
+            fillfieldvectorfromjavatocpp = env->GetStaticMethodID(javaClassToBeCalledByCpp, "fillfieldvectorfromjavatocpp", "(JJ)V");
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:399]: Shorter names will make it easier to read once rendered in Sphinx

File: docs/source/java/cdata.rst [hunk 13]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+            if ( fillfieldvectorfromjavatocpp != NULL ){
+                struct ArrowSchema arrowSchema;
+                struct ArrowArray arrowArray;
+                printf("\n<<<<< C++ to Java for Arrays >>>>>\n");
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:403]: Use std::cout in C++, not printf

File: docs/source/java/cdata.rst [hunk 14]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+                env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillfieldvectorfromjavatocpp, reinterpret_cast<uintptr_t>(&arrowSchema), reinterpret_cast<uintptr_t>(&arrowArray));
+                auto resultImportArray = arrow::ImportArray(&arrowArray, &arrowSchema);
+                std::shared_ptr<arrow::Array> array = resultImportArray.ValueOrDie();
+                auto int32_array = std::static_pointer_cast<arrow::Int32Array>(array);
+                const int32_t* data = int32_array->raw_values();
+                for (int j = 0; j < int32_array->length(); j++){
+                    printf("[C++ - side]: Data ImportArray - array[%d] = %d\n", j, data[j]);
+                }
+            } else {
+                printf("Problem to read method fillfieldvectorfromjavatocpp\n");
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:413]: std::abort or return EXIT_FAILURE

File: docs/source/java/cdata.rst [hunk 15]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+        jvm ->DestroyJavaVM();
+        return 0;
+    }
+
+CMakeLists.txt definition file:
+
+.. code-block:: xml
+
+    cmake_minimum_required(VERSION 3.19)
+    project(firstarrowcpp)
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:446]: name the project something related?

File: docs/source/java/cdata.rst [hunk 16]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+    find_package(JNI REQUIRED)
+    find_package(Arrow REQUIRED)
+    message(${ARROW_VERSION})
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:449]:
```suggestion
    message(STATUS "Arrow version: ${ARROW_VERSION}")
```

File: docs/source/java/cdata.rst [hunk 17]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+    message(${ARROW_FULL_SO_VERSION})
+    include_directories(${JNI_INCLUDE_DIRS}) #'jni.h' file not found
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:451]:
```suggestion
    include_directories(${JNI_INCLUDE_DIRS})
```

File: docs/source/java/cdata.rst [hunk 18]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+    set(CMAKE_CXX_STANDARD 14)
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:452]:
```suggestion
    set(CMAKE_CXX_STANDARD 11)
```
don't think we need C++14

File: docs/source/java/cdata.rst [hunk 19]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+    add_executable(${PROJECT_NAME} main.cpp)
+    target_link_libraries(firstarrowcpp PRIVATE arrow_shared)
+    target_link_libraries(firstarrowcpp PRIVATE ${JNI_LIBRARIES})
+
+**C++ Test**
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:457]:
```suggestion
**Result**
```

File: docs/source/java/cdata.rst [hunk 20]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+
+This is the result of the test:
+
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:460]:
```suggestion
```

File: docs/source/java/cdata.rst [hunk 21]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+            }
+            jmethodID fillvectorschemarootfromjavatocpp = NULL;
+            fillvectorschemarootfromjavatocpp = env->GetStaticMethodID(javaClassToBeCalledByCpp, "fillvectorschemarootfromjavatocpp", "(JJ)V");
+            if ( fillvectorschemarootfromjavatocpp != NULL ){
+                struct ArrowSchema arrowSchema;
+                struct ArrowArray arrowArray;
+                printf("\n<<<<< C++ to Java for RecordBatch >>>>>\n");
+                env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillvectorschemarootfromjavatocpp, static_cast<long>(reinterpret_cast<uintptr_t>(&arrowSchema)), static_cast<long>(reinterpret_cast<uintptr_t>(&arrowArray)));
+                auto resultImportVectorSchemaRoot = arrow::ImportRecordBatch(&arrowArray, &arrowSchema);
+                std::shared_ptr<arrow::RecordBatch> recordBatch = resultImportVectorSchemaRoot.ValueOrDie();
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:423]: recordBatch has a ToString

File: docs/source/java/cdata.rst [hunk 22]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+        if ( javaClassToBeCalledByCpp != NULL ) {
+            jmethodID fillfieldvectorfromjavatocpp = NULL;
+            fillfieldvectorfromjavatocpp = env->GetStaticMethodID(javaClassToBeCalledByCpp, "fillfieldvectorfromjavatocpp", "(JJ)V");
+            if ( fillfieldvectorfromjavatocpp != NULL ){
+                struct ArrowSchema arrowSchema;
+                struct ArrowArray arrowArray;
+                printf("\n<<<<< C++ to Java for Arrays >>>>>\n");
+                env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillfieldvectorfromjavatocpp, reinterpret_cast<uintptr_t>(&arrowSchema), reinterpret_cast<uintptr_t>(&arrowArray));
+                auto resultImportArray = arrow::ImportArray(&arrowArray, &arrowSchema);
+                std::shared_ptr<arrow::Array> array = resultImportArray.ValueOrDie();
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:406]: array has a ToString

File: docs/source/java/cdata.rst [hunk 23]
```diff
@@ -33,7 +33,10 @@ Python communication using the C Data Interface.
 Java to C++
 -----------
 
-Example: Share an Int64 array from C++ to Java:
+Share an Int64 array from C++ to Java
+=====================================
```
[the remark this answers is not in the document]
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Updated

File: docs/source/java/cdata.rst [hunk 24]
```diff
@@ -220,4 +223,259 @@ Let's create a Java class to test our bridge:
...
+            intVector.setSafe(0, 10);
+            intVector.setSafe(1, 20);
+            intVector.setSafe(2, 30);
+            VectorSchemaRoot root = new VectorSchemaRoot(Arrays.asList(intVector));
+            root.setRowCount(3);
+            System.out.println("[Java - side] VectorSchemaRoot: \n" + root.contentToTSVString());
+            return root;
+        }
+
+        public static void main(String[] args) {
```
[the remark this answers is not in the document]
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Deleted

File: docs/source/java/cdata.rst (hunk 3 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Changed

File: docs/source/java/cdata.rst (hunk 4 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Added

File: docs/source/java/cdata.rst (hunk 8 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Changed

File: docs/source/java/cdata.rst (hunk 11 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Reformat code added

File: docs/source/java/cdata.rst (hunk 13 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Changed

File: docs/source/java/cdata.rst (hunk 15 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Updated

File: docs/source/java/cdata.rst (hunk 14 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Added

File: docs/source/java/cdata.rst (hunk 21 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst:416]: Changed

File: docs/source/java/cdata.rst (hunk 22 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst:400]: Changed

File: docs/source/java/cdata.rst [hunk 25]
```diff
@@ -220,4 +221,252 @@ Let's create a Java class to test our bridge:
...
+                        </execution>
+                    </executions>
+                </plugin>
+            </plugins>
+        </build>
+    </project>
+
+.. code-block:: java
+
+    import org.apache.arrow.c.ArrowArray;
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:296]: nit, but everything is indented one space too many here

Comment (lidavidm) [reply docs/source/java/cdata.rst:296]: that's actually true of all the code blocks

File: docs/source/java/cdata.rst [hunk 26]
```diff
@@ -220,4 +221,252 @@ Let's create a Java class to test our bridge:
...
+        JavaVMOption options[2];
+        options[0].optionString = "-Djava.class.path=cpptojava.jar"; // java jar name
+        options[1].optionString = "-DXcheck:jni:pedantic";
+        vm_args.version = JNI_VERSION_1_8;
+        vm_args.nOptions = 2;
+        vm_args.options = options;
+        int status = JNI_CreateJavaVM(jvm, (void **) &env, &vm_args);
+        if (status < 0) {
+            std::cout << "\n<<<<< Unable to Launch JVM >>>>>\n" << std::endl;
+            std::abort();
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:378]: Requires `#include <cstdlib>`

File: docs/source/java/cdata.rst [hunk 27]
```diff
@@ -220,4 +221,252 @@ Let's create a Java class to test our bridge:
...
+
+    int main() {
+        JNIEnv *env;
+        JavaVM *jvm;
+        env = CreateVM(&jvm);
+        if (env == nullptr) return 1;
+        jclass javaClassToBeCalledByCpp = NULL;
+        javaClassToBeCalledByCpp = env->FindClass("ToBeCalledByCpp");
+        if (javaClassToBeCalledByCpp != NULL) {
+            jmethodID fillVector = NULL;
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:391]: Please update the other uses of NULL to be nullptr

File: docs/source/java/cdata.rst [hunk 28]
```diff
@@ -220,4 +221,252 @@ Let's create a Java class to test our bridge:
...
+CMakeLists.txt definition file:
+
+.. code-block:: xml
+
+    cmake_minimum_required(VERSION 3.19)
+    project(cdatacpptojava)
+    find_package(JNI REQUIRED)
+    find_package(Arrow REQUIRED)
+    message(STATUS "Arrow version: ${ARROW_VERSION}")
+    message(${ARROW_FULL_SO_VERSION})
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:439]: This needs to be updated like the line above

File: docs/source/java/cdata.rst [hunk 29]
```diff
@@ -220,4 +221,252 @@ Let's create a Java class to test our bridge:
...
+            std::abort();
+        }
+        return env;
+    }
+
+    int main() {
+        JNIEnv *env;
+        JavaVM *jvm;
+        env = CreateVM(&jvm);
+        if (env == nullptr) return 1;
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:387]: use EXIT_FAILURE here too as well

File: docs/source/java/cdata.rst (hunk 26 above)
Comment (lidavidm) [reply docs/source/java/cdata.rst:378]: Actually I suppose the `abort` isn't necessary, maybe just explicitly return `nullptr` and it'll exit in main()

File: docs/source/java/cdata.rst (hunk 27 above)
Comment (lidavidm) [reply docs/source/java/cdata.rst:391]: Though, why separate the declaration and assignment? Just do it in one line

File: docs/source/java/cdata.rst [hunk 30]
```diff
@@ -220,4 +221,252 @@ Let's create a Java class to test our bridge:
...
+        jclass javaClassToBeCalledByCpp = NULL;
+        javaClassToBeCalledByCpp = env->FindClass("ToBeCalledByCpp");
+        if (javaClassToBeCalledByCpp != NULL) {
+            jmethodID fillVector = NULL;
+            fillVector = env->GetStaticMethodID(javaClassToBeCalledByCpp, "fillVector", "(JJ)V");
+            if (fillVector != NULL) {
+                struct ArrowSchema arrowSchema;
+                struct ArrowArray arrowArray;
+                std::cout << "\n<<<<< C++ to Java for Arrays >>>>>\n" << std::endl;
+                env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillVector, reinterpret_cast<uintptr_t>(&arrowSchema),
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:397]: I would wrap lines to be a little shorter here if possible to make sure it's readable in Sphinx

File: docs/source/java/cdata.rst [hunk 31]
```diff
@@ -220,4 +221,252 @@ Let's create a Java class to test our bridge:
...
+                                          reinterpret_cast<uintptr_t>(&arrowArray));
+                auto resultImportArray = arrow::ImportArray(&arrowArray, &arrowSchema);
+                std::shared_ptr<arrow::Array> array = resultImportArray.ValueOrDie();
+                std::cout << "[C++] Array: " << array->ToString() << std::endl;
+            } else {
+                std::cout << "Problem to read method fillVector\n" << std::endl;
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:403]:
```suggestion
                std::cout << "Could not find fillVector method\n" << std::endl;
```

File: docs/source/java/cdata.rst [hunk 32]
```diff
@@ -220,4 +221,252 @@ Let's create a Java class to test our bridge:
...
+                struct ArrowArray arrowArray;
+                std::cout << "\n<<<<< C++ to Java for RecordBatch >>>>>\n" << std::endl;
+                env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillVectorSchemaRoot,
+                                          static_cast<long>(reinterpret_cast<uintptr_t>(&arrowSchema)),
+                                          static_cast<long>(reinterpret_cast<uintptr_t>(&arrowArray)));
+                auto resultImportVectorSchemaRoot = arrow::ImportRecordBatch(&arrowArray, &arrowSchema);
+                std::shared_ptr<arrow::RecordBatch> recordBatch = resultImportVectorSchemaRoot.ValueOrDie();
+                std::cout << "[C++] RecordBatch: " << recordBatch->ToString() << std::endl;
+            } else {
+                std::cout << "Problem to read method fillVectorSchemaRoot\n" << std::endl;
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:419]:
```suggestion
                std::cout << "Could not find fillVectorSchemaRoot method\n" << std::endl;
```

File: docs/source/java/cdata.rst [hunk 33]
```diff
@@ -220,4 +221,252 @@ Let's create a Java class to test our bridge:
...
+                return EXIT_FAILURE;
+            }
+        } else {
+            std::cout << "Problem to read class ToBeCalledByCpp\n" << std::endl;
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:423]:
```suggestion
            std::cout << "Could not find ToBeCalledByCpp class\n" << std::endl;
```

File: docs/source/java/cdata.rst [hunk 34]
```diff
@@ -220,4 +221,252 @@ Let's create a Java class to test our bridge:
...
+            return EXIT_FAILURE;
+        }
+        jvm->DestroyJavaVM();
+        return 0;
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:427]:
```suggestion
        return EXIT_SUCCESS;
```

File: docs/source/java/cdata.rst (hunk 27 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Updated

File: docs/source/java/cdata.rst (hunk 29 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Added

File: docs/source/java/cdata.rst (hunk 26 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Changed

File: docs/source/java/cdata.rst (hunk 30 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Updated

File: docs/source/java/cdata.rst (hunk 28 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Message not needed, deleted

File: docs/source/java/cdata.rst (hunk 25 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Changed

File: docs/source/java/cdata.rst [hunk 35]
```diff
@@ -91,133 +92,381 @@ without writing JNI bindings ourselves.
...
+   <project xmlns="http://maven.apache.org/POM/4.0.0"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+       <modelVersion>4.0.0</modelVersion>
+       <groupId>org.example</groupId>
+       <artifactId>cpptojava</artifactId>
+       <version>1.0-SNAPSHOT</version>
+       <properties>
+           <maven.compiler.source>11</maven.compiler.source>
+           <maven.compiler.target>11</maven.compiler.target>
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:243]: Any reason to use Java 11 here?

File: docs/source/java/cdata.rst [hunk 36]
```diff
@@ -91,133 +92,381 @@ without writing JNI bindings ourselves.
...
+           <arrow.version>8.0.0</arrow.version>
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:244]: Update to 9.0.0?

File: docs/source/java/cdata.rst [hunk 37]
```diff
@@ -91,133 +92,381 @@ without writing JNI bindings ourselves.
...
+       <modelVersion>4.0.0</modelVersion>
+
+       <groupId>org.example</groupId>
+       <artifactId>java-cdata-example</artifactId>
+       <version>1.0-SNAPSHOT</version>
+
+       <properties>
+           <maven.compiler.source>8</maven.compiler.source>
+           <maven.compiler.target>8</maven.compiler.target>
+           <arrow.version>8.0.0</arrow.version>
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:108]: Might as well update to 9.0.0 now

File: docs/source/java/cdata.rst [hunk 38]
```diff
@@ -91,133 +92,381 @@ without writing JNI bindings ourselves.
...
+       <version>1.0-SNAPSHOT</version>
+       <properties>
+           <maven.compiler.source>11</maven.compiler.source>
+           <maven.compiler.target>11</maven.compiler.target>
+           <arrow.version>8.0.0</arrow.version>
+       </properties>
+       <repositories>
+           <repository>
+               <id>arrow-nightly</id>
+               <url>https://nightlies.apache.org/arrow/java</url>
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:249]: No need to use nightlies right?

File: docs/source/java/cdata.rst [hunk 39]
```diff
@@ -91,133 +92,381 @@ without writing JNI bindings ourselves.
...
+                       <execution>
+                           <phase>package</phase>
+                           <goals>
+                               <goal>single</goal>
+                           </goals>
+                           <configuration>
+                               <archive>
+                                   <manifest>
+                                       <mainClass>
+                                           ToBeCalledByCpp
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:279]: Do we need the mainClass config?

File: docs/source/java/cdata.rst [hunk 40]
```diff
@@ -91,133 +92,381 @@ without writing JNI bindings ourselves.
...
+           }
+           jmethodID fillVectorSchemaRoot = env->GetStaticMethodID(javaClassToBeCalledByCpp,
+                                                                   "fillVectorSchemaRoot",
+                                                                   "(JJ)V");
+           if (fillVectorSchemaRoot != nullptr) {
+               struct ArrowSchema arrowSchema;
+               struct ArrowArray arrowArray;
+               std::cout << "\n<<<<< C++ to Java for RecordBatch >>>>>\n" << std::endl;
+               env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillVectorSchemaRoot,
+                                         static_cast<long>(reinterpret_cast<uintptr_t>(&arrowSchema)),
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:414]: Don't we need the `static_cast<long>` above as well then?

(Also, shouldn't it be `jlong`?)

Comment (davisusanibar) [reply docs/source/java/cdata.rst:414]: All of these kind of declaration work well (I don't know what of this should I use?)

```cpp
        env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillVector,
                                  static_cast<long>(reinterpret_cast<uintptr_t>(&arrowSchema)),
                                  static_cast<long>(reinterpret_cast<uintptr_t>(&arrowArray)));
        env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillVector,
                                  reinterpret_cast<uintptr_t>(&arrowSchema),
                                  reinterpret_cast<uintptr_t>(&arrowArray));
        env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillVector,
                                  (long)(&arrowSchema),
                                  (long)(&arrowArray));
        env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillVector,
                                  (jlong)(&arrowSchema),
                                  (jlong)(&arrowArray));
        env->CallStaticVoidMethod(javaClassToBeCalledByCpp, fillVector,
                                  (uintptr_t)(&arrowSchema),
                                  (uintptr_t)(&arrowArray));
```

File: docs/source/java/cdata.rst [hunk 41]
```diff
@@ -91,133 +92,381 @@ without writing JNI bindings ourselves.
...
+           std::cout << "Could not find ToBeCalledByCpp class\n" << std::endl;
+           return EXIT_FAILURE;
+       }
+       jvm->DestroyJavaVM();
+       return EXIT_SUCCESS;
+   }
+
+CMakeLists.txt definition file:
+
+.. code-block:: xml
```
Comment (pitrou) [inline docs/source/java/cdata.rst:433]:
```suggestion
.. code-block:: cmake
```

File: docs/source/java/cdata.rst [hunk 42]
```diff
@@ -91,133 +92,381 @@ without writing JNI bindings ourselves.
...
+   message(STATUS "Arrow version: ${ARROW_VERSION}")
+   include_directories(${JNI_INCLUDE_DIRS})
+   set(CMAKE_CXX_STANDARD 11)
+   add_executable(${PROJECT_NAME} main.cpp)
+   target_link_libraries(cdatacpptojava PRIVATE arrow_shared)
+   target_link_libraries(cdatacpptojava PRIVATE ${JNI_LIBRARIES})
+
+**Result**
 
 .. code-block:: shell
```
Comment (pitrou) [inline docs/source/java/cdata.rst:448]: This isn't shell code, and the highlighting would look funny.
```suggestion
.. code-block:: text
```

File: docs/source/java/cdata.rst [hunk 43]
```diff
@@ -91,133 +92,381 @@ without writing JNI bindings ourselves.
...
+       final static BufferAllocator allocator = new RootAllocator();
+
+       public static void fillVector(long schemaAddress, long arrayAddress){
+           try (ArrowArray arrow_array = ArrowArray.wrap(arrayAddress);
+                ArrowSchema arrow_schema = ArrowSchema.wrap(schemaAddress) ) {
+               Data.exportVector(allocator, populateFieldVectorToExport(), null, arrow_array, arrow_schema);
+           }
+       }
+
+       public static FieldVector populateFieldVectorToExport(){
```
Comment (pitrou) [inline docs/source/java/cdata.rst:317]: Shouldn't you make this private?

File: docs/source/java/cdata.rst [hunk 44]
```diff
@@ -91,133 +92,381 @@ without writing JNI bindings ourselves.
...
+       }
+
+       public static void fillVectorSchemaRoot(long schemaAddress, long arrayAddress){
+           try (ArrowArray arrow_array = ArrowArray.wrap(arrayAddress);
+                ArrowSchema arrow_schema = ArrowSchema.wrap(schemaAddress) ) {
+               Data.exportVectorSchemaRoot(allocator, populateVectorSchemaRootToExport(), null, arrow_array, arrow_schema);
+           }
+       }
+
+       public static VectorSchemaRoot populateVectorSchemaRootToExport(){
```
Comment (pitrou) [inline docs/source/java/cdata.rst:335]: private as well?

File: docs/source/java/cdata.rst [hunk 45]
```diff
@@ -91,133 +92,381 @@ without writing JNI bindings ourselves.
...
+   import org.apache.arrow.vector.FieldVector;
+   import org.apache.arrow.vector.IntVector;
+   import org.apache.arrow.vector.VectorSchemaRoot;
+
+   import java.util.Arrays;
+
+   public class ToBeCalledByCpp {
+       final static BufferAllocator allocator = new RootAllocator();
+
+       public static void fillVector(long schemaAddress, long arrayAddress){
```
Comment (pitrou) [inline docs/source/java/cdata.rst:310]: Can you add small docstring describing the two public methods?

File: docs/source/java/cdata.rst (hunk 40 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Thank you

File: docs/source/java/cdata.rst (hunk 38 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Deleted

File: docs/source/java/cdata.rst (hunk 39 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Deleted

File: docs/source/java/cdata.rst (hunk 43 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Changed

File: docs/source/java/cdata.rst (hunk 44 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Changed

File: docs/source/java/cdata.rst (hunk 35 above)
Comment (davisusanibar) [reply docs/source/java/cdata.rst]: Deleted

File: docs/source/java/cdata.rst [hunk 46]
```diff
@@ -91,133 +92,379 @@ without writing JNI bindings ourselves.
...
+   import org.apache.arrow.vector.IntVector;
+   import org.apache.arrow.vector.VectorSchemaRoot;
+
+   import java.util.Arrays;
+
+   public class ToBeCalledByCpp {
+       final static BufferAllocator allocator = new RootAllocator();
+
+       /**
+        * Create a {@link FieldVector} and export this to an existing memory address
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:298]:
```suggestion
        * Create a {@link FieldVector} and export it via the C Data Interface
```

File: docs/source/java/cdata.rst [hunk 47]
```diff
@@ -91,133 +92,379 @@ without writing JNI bindings ourselves.
...
+        */
+       public static void fillVector(long schemaAddress, long arrayAddress){
+           try (ArrowArray arrow_array = ArrowArray.wrap(arrayAddress);
+                ArrowSchema arrow_schema = ArrowSchema.wrap(schemaAddress) ) {
+               Data.exportVector(allocator, populateFieldVectorToExport(), null, arrow_array, arrow_schema);
+           }
+       }
+
+       /**
+        * Create a {@link VectorSchemaRoot} and export this to an existing memory address
```
Comment (lidavidm) [inline docs/source/java/cdata.rst:310]:
```suggestion
        * Create a {@link VectorSchemaRoot} and export it via the C Data Interface
```

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
Lint issues are addressed by https://github.com/apache/arrow/pull/13809

**Pull request merged**